File: backend/mcp-code-intelligence-kotlin/src/main/kotlin/com/codeintel/analyzers/complexity/GradeAssigner.kt
/**
 * KSA-161: Grade assignment based on cyclomatic complexity thresholds.
 */
package com.codeintel.analyzers.complexity

class GradeAssigner(thresholds: GradeThresholds? = null) {
    private val thresholds: GradeThresholds = thresholds ?: GradeThresholds()

    fun assignGrade(cc: Int): Grade = when {
        cc <= thresholds.a -> Grade.A
        cc <= thresholds.b -> Grade.B
        cc <= thresholds.c -> Grade.C
        cc <= thresholds.d -> Grade.D
        else -> Grade.F
    }

    fun getThresholds(): GradeThresholds = thresholds.copy()
}

File: backend/src/engine/analyzers/complexity/GradeAssigner.ts
/**
 * KSA-161: Grade assignment based on cyclomatic complexity thresholds.
 */

import type { Grade, GradeThresholds } from './types.js';

const DEFAULT_THRESHOLDS: GradeThresholds = {
  A: 5,
  B: 10,
  C: 20,
  D: 50,
};

export class GradeAssigner {
  private thresholds: GradeThresholds;

  constructor(thresholds?: Partial<GradeThresholds>) {
    this.thresholds = { ...DEFAULT_THRESHOLDS, ...thresholds };
  }

  /** Assign a letter grade based on cyclomatic complexity score. */
  assignGrade(cc: number): Grade {
    if (cc <= this.thresholds.A) return 'A';
    if (cc <= this.thresholds.B) return 'B';
    if (cc <= this.thresholds.C) return 'C';
    if (cc <= this.thresholds.D) return 'D';
    return 'F';
  }

  /** Get the current thresholds. */
  getThresholds(): GradeThresholds {
    return { ...this.thresholds };
  }
}
